File: app/services/content_service/content_service_stub.py
from typing import AsyncIterator, Optional

import grpc
import grpc.aio
from oink_judge import content_service_pb2 as pb2
from oink_judge import content_service_pb2_grpc as pb2_grpc
from oink_judge.pybind11_content_service import get_content_storage_stub_type

import app.utils.grpc  # noqa: F401 — registers channel/credential factory types
from app.factory import ParameterizedTypeFactory, normalize_argument, parse_parameters
from app.utils.grpc import ChannelFactory

CHUNK_SIZE = 64 * 1024  # 64 KB


class ContentServiceStubFactory(ParameterizedTypeFactory["ContentServiceStub"]): ...


def get_content_storage_stub() -> Optional["ContentServiceStub"]:
    stub_type = get_content_storage_stub_type()
    if stub_type is None:
        return None
    return ContentServiceStubFactory.instance().create(stub_type)


def register_content_service_channel_stub_type() -> None:
    def channel_stub(params: str) -> "ContentServiceStub":
        parts = parse_parameters(params, ",")
        if len(parts) != 1:
            raise ValueError(
                "Expected exactly one parameter for content_service_stub: channel type"
            )
        channel_spec = normalize_argument(parts[0], True)
        channel = ChannelFactory.instance().create(channel_spec)
        return ContentServiceStub(channel)

    ContentServiceStubFactory.instance().register_type(
        "content_service_stub", channel_stub
    )


register_content_service_channel_stub_type()


class ContentServiceStub:
    _stub: pb2_grpc.ContentServiceStub

    def __init__(self, channel: grpc.aio.Channel) -> None:
        self._stub = pb2_grpc.ContentServiceStub(channel)

    async def get_manifest(self, content_type: str, content_id: str) -> dict:
        import json

        request = pb2.GetManifestRequest(
            content_type=content_type,
            content_id=content_id,
        )
        parts: list[str] = []
        async for response in self._stub.GetManifest(request):
            parts.append(response.manifest)

        return json.loads("".join(parts))

    async def get_file(
        self, content_type: str, content_id: str, file_path: str
    ) -> bytes:
        request = pb2.GetFileRequest(
            content_type=content_type,
            content_id=content_id,
            file_path=file_path,
        )
        chunks: list[bytes] = []
        async for response in self._stub.GetFile(request):
            chunks.append(response.file_chunk.data)

        return b"".join(chunks)

    async def create_file(
        self,
        content_type: str,
        content_id: str,
        file_path: str,
        file_content: bytes,
    ) -> None:
        await self._stub.CreateFile(
            self._file_upload_requests(
                pb2.CreateFileRequest,
                content_type,
                content_id,
                file_path,
                file_content,
            )
        )

    async def update_file(
        self,
        content_type: str,
        content_id: str,
        file_path: str,
        file_content: bytes,
    ) -> None:
        await self._stub.UpdateFile(
            self._file_upload_requests(
                pb2.UpdateFileRequest,
                content_type,
                content_id,
                file_path,
                file_content,
            )
        )

    async def delete_file(
        self, content_type: str, content_id: str, file_path: str
    ) -> None:
        request = pb2.DeleteFileRequest(
            content_type=content_type,
            content_id=content_id,
            file_path=file_path,
        )
        await self._stub.DeleteFile(request)

    @staticmethod
    async def _file_upload_requests(
        request_cls,  # CreateFileRequest | UpdateFileRequest
        content_type: str,
        content_id: str,
        file_path: str,
        file_content: bytes,
    ) -> AsyncIterator:
        # First message: metadata
        info = pb2.FileInfo(
            content_type=content_type,
            content_id=content_id,
            file_path=file_path,
        )
        yield request_cls(file_info=info)

        # Subsequent messages: data chunks
        for offset in range(0, len(file_content), CHUNK_SIZE):
            chunk = pb2.FileChunk(data=file_content[offset : offset + CHUNK_SIZE])
            yield request_cls(file_chunk=chunk)

File: modules/content_service/src/client/content_service_stub.cpp
#include "oink_judge/content_service/client/content_service_stub.h"

#include "agrpc/grpc_context.hpp"
#include "oink_judge/factory/parameterized_type_factory.hpp"
#include "tl/expected.hpp"

#include <oink_judge/utils/grpc/factories.hpp>

#include <agrpc/client_rpc.hpp>
#include <boost/asio.hpp>
#include <grpcpp/client_context.h>
#include <grpcpp/grpcpp.h>

#include <string>

namespace oink_judge::content_service {

namespace {

using GetManifestRPC = agrpc::ClientRPC<&ContentService::Stub::PrepareAsyncGetManifest>;
using GetFileRPC = agrpc::ClientRPC<&ContentService::Stub::PrepareAsyncGetFile>;
using CreateFileRPC = agrpc::ClientRPC<&ContentService::Stub::PrepareAsyncCreateFile>;
using UpdateFileRPC = agrpc::ClientRPC<&ContentService::Stub::PrepareAsyncUpdateFile>;
using DeleteFileRPC = agrpc::ClientRPC<&ContentService::Stub::PrepareAsyncDeleteFile>;

} // namespace

auto ContentServiceChannelStub::getManifest(std::string content_type, std::string content_id)
    -> awaitable<tl::expected<json, grpc::Status>> {
    auto& grpc_context = static_cast<agrpc::GrpcContext&>((co_await boost::asio::this_coro::executor).context()); // NOLINT

    GetManifestRequest request;
    request.set_content_type(std::move(content_type));
    request.set_content_id(std::move(content_id));

    GetManifestResponse response;

    auto rpc = GetManifestRPC(grpc_context);
    co_await rpc.start(*stub_, request, boost::asio::use_awaitable);

    std::ostringstream manifest_stream;

    while (co_await rpc.read(response, boost::asio::use_awaitable)) {
        manifest_stream << response.manifest();
    }

    grpc::Status status = co_await rpc.finish(boost::asio::use_awaitable);

    if (!status.ok()) {
        co_return tl::unexpected(status);
    }
    json manifest_json = json::parse(manifest_stream.str());
    co_return manifest_json;
}

auto ContentServiceChannelStub::getFile(std::string content_type, std::string content_id, std::string file_path)
    -> awaitable<tl::expected<std::string, grpc::Status>> {
    auto& grpc_context = static_cast<agrpc::GrpcContext&>((co_await boost::asio::this_coro::executor).context()); // NOLINT

    GetFileRequest request;
    request.set_content_type(std::move(content_type));
    request.set_content_id(std::move(content_id));
    request.set_file_path(std::move(file_path));

    GetFileResponse response;

    auto rpc = GetFileRPC(grpc_context);
    co_await rpc.start(*stub_, request, boost::asio::use_awaitable);

    std::ostringstream file_content_stream(std::ios::binary);

    while (co_await rpc.read(response, boost::asio::use_awaitable)) {
        file_content_stream << response.file_chunk().data();
    }

    grpc::Status status = co_await rpc.finish(boost::asio::use_awaitable);

    if (!status.ok()) {
        co_return tl::unexpected(status);
    }
    co_return file_content_stream.str();
}

auto ContentServiceChannelStub::createFile(std::string content_type, std::string content_id, std::string file_path,
                                           std::string file_content) -> awaitable<tl::expected<void, grpc::Status>> {
    auto& grpc_context = static_cast<agrpc::GrpcContext&>((co_await boost::asio::this_coro::executor).context()); // NOLINT

    CreateFileRequest info;
    google::protobuf::Empty response;

    auto rpc = CreateFileRPC(grpc_context);

    auto* file_info = info.mutable_file_info();
    file_info->set_content_type(std::move(content_type));
    file_info->set_content_id(std::move(content_id));
    file_info->set_file_path(std::move(file_path));

    co_await rpc.start(*stub_, response, boost::asio::use_awaitable);

    co_await rpc.write(info, boost::asio::use_awaitable);

    constexpr auto CHUNK_SIZE = static_cast<size_t>(64 * 1024); // 64KB
    std::vector<char> buffer(CHUNK_SIZE);
    std::istringstream content_stream(std::move(file_content), std::ios::binary);

    while (content_stream) {
        content_stream.read(buffer.data(), static_cast<std::streamsize>(buffer.size()));
        std::streamsize bytes_read = content_stream.gcount();
        if (bytes_read == 0) {
            break;
        }
        if (bytes_read > 0) {
            CreateFileRequest chunk_request;
            auto* chunk = chunk_request.mutable_file_chunk();
            chunk->set_data(buffer.data(), static_cast<size_t>(bytes_read));
            co_await rpc.write(chunk_request, boost::asio::use_awaitable);
        }
    }

    grpc::Status status = co_await rpc.finish(boost::asio::use_awaitable);

    if (!status.ok()) {
        co_return tl::unexpected(status);
    }
    co_return tl::expected<void, grpc::Status>{};
}

auto ContentServiceChannelStub::updateFile(std::string content_type, std::string content_id, std::string file_path,
                                           std::string file_content) -> awaitable<tl::expected<void, grpc::Status>> {
    auto& grpc_context = static_cast<agrpc::GrpcContext&>((co_await boost::asio::this_coro::executor).context()); // NOLINT

    UpdateFileRequest info;
    google::protobuf::Empty response;

    auto* file_info = info.mutable_file_info();
    file_info->set_content_type(std::move(content_type));
    file_info->set_content_id(std::move(content_id));
    file_info->set_file_path(std::move(file_path));

    auto rpc = UpdateFileRPC(grpc_context);

    co_await rpc.start(*stub_, response, boost::asio::use_awaitable);

    co_await rpc.write(info, boost::asio::use_awaitable);

    constexpr auto CHUNK_SIZE = static_cast<size_t>(64 * 1024); // 64KB
    std::vector<char> buffer(CHUNK_SIZE);
    std::istringstream content_stream(std::move(file_content), std::ios::binary);

    while (content_stream) {
        content_stream.read(buffer.data(), static_cast<std::streamsize>(buffer.size()));
        std::streamsize bytes_read = content_stream.gcount();
        if (bytes_read == 0) {
            break;
        }
        if (bytes_read > 0) {
            UpdateFileRequest chunk_request;
            auto* chunk = chunk_request.mutable_file_chunk();
            chunk->set_data(buffer.data(), static_cast<size_t>(bytes_read));
            co_await rpc.write(chunk_request, boost::asio::use_awaitable);
        }
    }

    grpc::Status status = co_await rpc.finish(boost::asio::use_awaitable);

    if (!status.ok()) {
        co_return tl::unexpected(status);
    }
    co_return tl::expected<void, grpc::Status>{};
}

auto ContentServiceChannelStub::deleteFile(std::string content_type, std::string content_id, std::string file_path)
    -> awaitable<tl::expected<void, grpc::Status>> {
    auto& grpc_context = static_cast<agrpc::GrpcContext&>((co_await boost::asio::this_coro::executor).context()); // NOLINT

    DeleteFileRequest request;
    request.set_content_type(std::move(content_type));
    request.set_content_id(std::move(content_id));
    request.set_file_path(std::move(file_path));

    google::protobuf::Empty response;
    grpc::ClientContext context;
    grpc::Status status =
        co_await DeleteFileRPC::request(grpc_context, *stub_, context, request, response, boost::asio::use_awaitable);

    if (!status.ok()) {
        co_return tl::unexpected(status);
    }
    co_return tl::expected<void, grpc::Status>{};
}

ContentServiceChannelStub::ContentServiceChannelStub(std::shared_ptr<grpc::Channel> channel) : channel_(std::move(channel)) {
    stub_ = ContentService::NewStub(channel_);
}

auto registerContentServiceChannelStubType() -> void {
    ContentServiceStubFactory::instance().registerType(
        ContentServiceChannelStub::REGISTERED_NAME, [](const std::string& param) -> std::unique_ptr<ContentServiceChannelStub> {
            std::vector<std::string> parts = factory::parseParameters(param, ",");
            if (parts.size() != 1) {
                throw std::invalid_argument("Expected exactly one parameter for channel_stub: channel type");
            }
            std::string channel_type = factory::normalizeArgument(parts[0], true);
            auto channel = utils::grpc::ChannelFactory::instance().create(channel_type);
            return std::make_unique<ContentServiceChannelStub>(std::move(channel));
        });
}

} // namespace oink_judge::content_service
